oauth2_provider/views/oidc.py

```diff
@@ -120,3 +147,246 @@ def _create_userinfo_response(self, request):
+            )
+        except OIDCError as error:
+            return self.error_response(error)
+
+        if not prompt:
+            return self.do_logout(application, redirect_uri, state)
+
+        self.oidc_data = {
+            "id_token_hint": id_token_hint,
+            "client_id": client_id,
+            "post_logout_redirect_uri": post_logout_redirect_uri,
+            "state": state,
+        }
+        form = self.get_form(self.get_form_class())
+        kwargs["form"] = form
+        if application:
+            kwargs["application"] = application
+
+        return self.render_to_response(self.get_context_data(**kwargs))
+
+    def form_valid(self, form):
+        id_token_hint = form.cleaned_data.get("id_token_hint")
+        client_id = form.cleaned_data.get("client_id")
+        post_logout_redirect_uri = form.cleaned_data.get("post_logout_redirect_uri")
+        state = form.cleaned_data.get("state")
+
+        try:
+            prompt, (redirect_uri, application) = validate_logout_request(
+                request=self.request,
+                id_token_hint=id_token_hint,
+                client_id=client_id,
+                post_logout_redirect_uri=post_logout_redirect_uri,
+            )
+
+            if not prompt or form.cleaned_data.get("allow"):
+                return self.do_logout(application, redirect_uri, state)
+            else:
+                raise LogoutDenied()
+
+        except OIDCError as error:
+            return self.error_response(error)
+
+    def do_logout(self, application=None, post_logout_redirect_uri=None, state=None):
+        # Delete Access Tokens
+        if oauth2_settings.OIDC_RP_INITIATED_LOGOUT_DELETE_TOKENS:
+            AccessToken = get_access_token_model()
+            RefreshToken = get_refresh_token_model()
+            access_tokens_to_delete = AccessToken.objects.filter(
+                user=self.request.user,
+                application__client_type__in=self.token_deletion_client_types,
+                application__authorization_grant_type__in=self.token_deletion_grant_types,
+            )
+            # This queryset has to be evaluated eagerly. The queryset would be empty with lazy evaluation
+            # because `access_tokens_to_delete` represents an empty queryset once `refresh_tokens_to_delete`
+            # is evaluated as all AccessTokens have been deleted.
+            refresh_tokens_to_delete = list(
+                RefreshToken.objects.filter(access_token__in=access_tokens_to_delete)
+            )
+            for token in access_tokens_to_delete:
+                # Delete the token and its corresponding refresh and IDTokens.
+                if token.id_token:
+                    token.id_token.revoke()
+                token.revoke()
+            for refresh_token in refresh_tokens_to_delete:
+                refresh_token.revoke()
+        # Logout in Django
+        logout(self.request)
```

Comment:
There's an issue here if the RP sent the request from it's server.

Django logout won't be able to remove the session created in the login view which will allow the user to login again from RP without entering their credentials.
 
We can clear the sessions with the following code (not really satisfying, but it works)
```
[
    s.delete()
    for s in Session.objects.filter(expire_date__gte=timezone.now())
    if s.get_decoded().get("_auth_user_id") == str(id_token.user_id)
]
```

Another way might be to link the Session to one of the tokens so that we can retrieve them.

Reply:
This also raises the question of `REFRESH_TOKEN_EXPIRE_SECONDS` vs `SESSION_COOKIE_AGE` : 
If Django session bigger than `REFRESH_TOKEN_EXPIRE_SECONDS` then we can be in the case where the tokens were deleted with `cleartokens` management command, but the session still exists.

Reply:
Although I don't want to cater to the use case of a back channel logout, deleting all sessions might be usefull. It is a shame that Django only allows such a hacky way to achieve this. I might add this together with a setting to enable/disable it.

Reply:
@tonial per https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout

> An RP requests that the OP log out the End-User by redirecting the **End-User's User Agent** to the OP's Logout Endpoint.

RP initiated logout does not allow an RP's server will make a direct call to the end_session_endpoint on behalf the user. The scenario you are describing is outside of the bounds of protocol and not applicable.

Reply:
@tonial also thank your for taking the time to review. The catch on the token parsing was particularly nice.